pkg/storegateway/series_chunks.go

```diff
@@ -365,24 +387,69 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
 
 			err := c.chunkReaders.addLoad(chunk.blockID, chunk.ref, i, j)
 			if err != nil {
-				c.err = errors.Wrap(err, "preloading chunks")
-				return false
+				return nil, false, errors.Wrap(err, "preloading chunks")
 			}
 		}
 	}
 
 	// Create a batched memory pool that can be released all at once.
 	chunksPool := pool.NewSafeSlabPool[byte](chunkBytesSlicePool, chunkBytesSlabSize)
 
-	err := c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
+	err = c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
 	if err != nil {
-		c.err = errors.Wrap(err, "loading chunks")
-		return false
+		return nil, false, errors.Wrap(err, "loading chunks")
 	}
 
 	nextSet.chunksReleaser = chunksPool
-	c.current = nextSet
-	return true
+
+	if c.ignoreNativeHistograms {
+		nextSet.series = filterNativeHistogramChunks(nextSet.series)
+		if len(nextSet.series) == 0 {
+			nextSet.release()
+			return nil, true, nil
+		}
+	}
+
+	return &nextSet, false, nil
+}
+
+func filterNativeHistogramChunks(seriesEntries []seriesEntry) []seriesEntry {
+	seriesWriteIdx := 0
+	for i, series := range seriesEntries {
+		chksWriteIdx := 0
+		for j, chk := range series.chks {
+			// The bucketChunkReader will skip populating native histogram chunks when ignoreNativeHistograms is true,
+			// leaving chk.Raw as nil.
+			if chk.Raw == nil {
+				chk.Reset()
```

Comment:
Why calling Reset() here? I don't think it's necessary. Could you explain why is it required?

Reply:
Yep! Same reason as here: https://github.com/grafana/mimir/pull/4059#discussion_r1093392707